"""
SMS gRPCæœåŠ¡å™¨å®ç° - æ”¯æŒå¤šè°ƒåˆ¶è§£è°ƒå™¨ç®¡ç†ï¼Œä¼˜åŒ–é”å®šæ—¶é—´
"""
import json
import time
import uuid
from typing import Optional, List
import grpc
from loguru import logger

from . import sms_pb2, sms_pb2_grpc
from src.common.serial_manager import SerialManager


class SMSService(sms_pb2_grpc.SMSServiceServicer):
    """SMSæœåŠ¡å®ç°"""

    def __init__(self, serial_manager: SerialManager):
        self.serial_manager = serial_manager
        self._send_timeout = 30  # å‘é€çŸ­ä¿¡çš„è¶…æ—¶æ—¶é—´ï¼ˆç§’ï¼‰
        self._lock_timeout = 25  # è·å–é”çš„è¶…æ—¶æ—¶é—´ï¼ˆç§’ï¼‰

    async def SendSMS(self, request, context) -> sms_pb2.SendSMSResponse:
        """å‘é€å•æ¡çŸ­ä¿¡"""
        logger.info(f"ğŸ“¨ å‘é€çŸ­ä¿¡: {request.phone_number}")
        start_time = time.time()

        try:
            # é€šè¿‡ä¸²å£ç®¡ç†å™¨å‘é€çŸ­ä¿¡
            success, message, modem_port = await self.serial_manager.send_sms(
                phone_number=request.phone_number,
                content=request.content
            )

            elapsed_time = time.time() - start_time

            # æ„å»ºå“åº”æ•°æ®
            response_data = {
                "message_id": str(uuid.uuid4()),
                "timestamp": time.time(),
                "phone_number": request.phone_number,
                "content_length": len(request.content),
                "success": success,
                "message": message,
                "elapsed_time": round(elapsed_time, 2),
                "modem_port": modem_port if modem_port else "unknown"
            }

            # æ·»åŠ è¯·æ±‚å…ƒæ•°æ®
            if request.metadata:
                response_data["metadata"] = dict(request.metadata)

            if request.sender_id:
                response_data["sender_id"] = request.sender_id

            if request.delivery_report:
                response_data["delivery_report"] = True

            status_code = 200 if success else 500

            # æ ¹æ®å¤„ç†æ—¶é—´è®°å½•æ—¥å¿—
            if elapsed_time > 20:
                logger.warning(f"âš ï¸ çŸ­ä¿¡å‘é€æ—¶é—´è¾ƒé•¿: {elapsed_time:.1f}ç§’ï¼Œä½¿ç”¨è°ƒåˆ¶è§£è°ƒå™¨: {modem_port}")
            elif elapsed_time > 10:
                logger.info(f"ğŸ“Š çŸ­ä¿¡å‘é€æ—¶é—´: {elapsed_time:.1f}ç§’ï¼Œä½¿ç”¨è°ƒåˆ¶è§£è°ƒå™¨: {modem_port}")
            else:
                logger.info(f"âœ… çŸ­ä¿¡å‘é€å®Œæˆ: {elapsed_time:.1f}ç§’ï¼Œä½¿ç”¨è°ƒåˆ¶è§£è°ƒå™¨: {modem_port}")

            return sms_pb2.SendSMSResponse(
                status=status_code,
                message=message,
                data=json.dumps(response_data, ensure_ascii=False)
            )

        except asyncio.TimeoutError:
            elapsed_time = time.time() - start_time
            logger.error(f"â° çŸ­ä¿¡å‘é€è¶…æ—¶: {request.phone_number} ({elapsed_time:.1f}ç§’)")
            error_data = {
                "error": "å‘é€è¶…æ—¶",
                "timestamp": time.time(),
                "phone_number": request.phone_number,
                "elapsed_time": round(elapsed_time, 2)
            }
            return sms_pb2.SendSMSResponse(
                status=504,
                message="çŸ­ä¿¡å‘é€è¶…æ—¶",
                data=json.dumps(error_data, ensure_ascii=False)
            )

        except Exception as e:
            elapsed_time = time.time() - start_time
            logger.error(f"çŸ­ä¿¡å‘é€å¤±è´¥: {request.phone_number} - {e} ({elapsed_time:.1f}ç§’)")
            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "phone_number": request.phone_number,
                "elapsed_time": round(elapsed_time, 2)
            }
            return sms_pb2.SendSMSResponse(
                status=500,
                message=f"å†…éƒ¨æœåŠ¡å™¨é”™è¯¯: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def SendBatchSMS(self, request, context) -> sms_pb2.SendBatchSMSResponse:
        """æ‰¹é‡å‘é€çŸ­ä¿¡"""
        logger.info(f"ğŸ“¦ æ‰¹é‡å‘é€çŸ­ä¿¡ï¼Œæ•°é‡: {len(request.phone_numbers)}")
        batch_start_time = time.time()

        results: List[dict] = []
        success_count = 0
        failed_count = 0
        modem_usage = {}

        try:
            # ä¸ºæ‰¹é‡å‘é€è®¾ç½®è¶…æ—¶
            batch_timeout = self._send_timeout * len(request.phone_numbers)

            for i, phone_number in enumerate(request.phone_numbers):
                try:
                    logger.info(f"   [{i+1}/{len(request.phone_numbers)}] å‘é€åˆ°: {phone_number}")

                    # é€šè¿‡ä¸²å£ç®¡ç†å™¨å‘é€çŸ­ä¿¡
                    success, message, modem_port = await self.serial_manager.send_sms(
                        phone_number=phone_number,
                        content=request.content
                    )

                    # è®°å½•è°ƒåˆ¶è§£è°ƒå™¨ä½¿ç”¨æƒ…å†µ
                    if modem_port:
                        modem_usage[modem_port] = modem_usage.get(modem_port, 0) + 1

                    # è®°å½•ç»“æœ
                    result_data = {
                        "message_id": str(uuid.uuid4()),
                        "phone_number": phone_number,
                        "status": 200 if success else 500,
                        "message": message,
                        "timestamp": time.time(),
                        "success": success
                    }

                    if modem_port:
                        result_data["modem_port"] = modem_port

                    results.append(result_data)

                    if success:
                        success_count += 1
                        logger.info(f"   âœ… ç¬¬ {i+1} æ¡å‘é€æˆåŠŸ")
                    else:
                        failed_count += 1
                        logger.error(f"   âŒ ç¬¬ {i+1} æ¡å‘é€å¤±è´¥: {message}")

                except asyncio.TimeoutError:
                    logger.error(f"   â° ç¬¬ {i+1} æ¡å‘é€è¶…æ—¶")
                    results.append({
                        "phone_number": phone_number,
                        "status": 504,
                        "message": "å‘é€è¶…æ—¶",
                        "timestamp": time.time(),
                        "error": True
                    })
                    failed_count += 1

                except Exception as e:
                    logger.error(f"æ‰¹é‡å‘é€å¤±è´¥ - {phone_number}: {e}")
                    results.append({
                        "phone_number": phone_number,
                        "status": 500,
                        "message": str(e),
                        "timestamp": time.time(),
                        "error": True
                    })
                    failed_count += 1

            # æ„å»ºæ‰¹é‡å“åº”æ•°æ®
            batch_elapsed_time = time.time() - batch_start_time
            batch_data = {
                "batch_id": str(uuid.uuid4()),
                "timestamp": time.time(),
                "total_count": len(request.phone_numbers),
                "success_count": success_count,
                "failed_count": failed_count,
                "results": results,
                "content": request.content,
                "content_length": len(request.content),
                "batch_elapsed_time": round(batch_elapsed_time, 2),
                "modem_usage": modem_usage
            }

            # æ·»åŠ è¯·æ±‚å…ƒæ•°æ®
            if request.metadata:
                batch_data["metadata"] = dict(request.metadata)

            if request.sender_id:
                batch_data["sender_id"] = request.sender_id

            if request.delivery_report:
                batch_data["delivery_report"] = True

            overall_status = 200 if success_count > 0 else 500
            overall_message = f"æ‰¹é‡å‘é€å®Œæˆ ({batch_elapsed_time:.1f}ç§’)ï¼ŒæˆåŠŸ {success_count} æ¡ï¼Œå¤±è´¥ {failed_count} æ¡"

            # è®°å½•è°ƒåˆ¶è§£è°ƒå™¨ä½¿ç”¨ç»Ÿè®¡
            if modem_usage:
                modem_stats = ", ".join([f"{k}: {v}æ¡" for k, v in modem_usage.items()])
                logger.info(f"ğŸ“Š è°ƒåˆ¶è§£è°ƒå™¨ä½¿ç”¨ç»Ÿè®¡: {modem_stats}")

            return sms_pb2.SendBatchSMSResponse(
                status=overall_status,
                message=overall_message,
                data=json.dumps(batch_data, ensure_ascii=False)
            )

        except Exception as e:
            batch_elapsed_time = time.time() - batch_start_time
            logger.error(f"æ‰¹é‡å‘é€å¤„ç†å¤±è´¥ ({batch_elapsed_time:.1f}ç§’): {e}")
            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "phone_numbers_count": len(request.phone_numbers),
                "batch_elapsed_time": round(batch_elapsed_time, 2)
            }
            return sms_pb2.SendBatchSMSResponse(
                status=500,
                message=f"æ‰¹é‡å‘é€å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def HealthCheck(self, request, context) -> sms_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥ - æä¾›è¯¦ç»†çŠ¶æ€ä¿¡æ¯"""
        try:
            if not self.serial_manager:
                return sms_pb2.HealthCheckResponse(
                    status=503,
                    message="æœåŠ¡ä¸å¥åº·: ä¸²å£ç®¡ç†å™¨æœªåˆå§‹åŒ–",
                    data=json.dumps({
                        "timestamp": time.time(),
                        "service_ready": False,
                        "error": "serial_manager_not_initialized",
                        "details": "Serial manager is not initialized"
                    }, ensure_ascii=False)
                )

            # è·å–ä¸²å£ç®¡ç†å™¨çŠ¶æ€
            health_status = await self.serial_manager.get_health_status()

            # æµ‹è¯•è¿æ¥ï¼ˆä½¿ç”¨è¾ƒçŸ­çš„è¶…æ—¶ï¼‰
            connected = await self.serial_manager.test_all_connections()

            # åˆ†æè°ƒåˆ¶è§£è°ƒå™¨çŠ¶æ€
            available_modems = []
            busy_modems = []
            offline_modems = []

            for modem in health_status["modems"]:
                if modem["is_available"]:
                    if modem["in_use"]:
                        busy_modems.append(modem)
                    else:
                        available_modems.append(modem)
                else:
                    offline_modems.append(modem)

            # æ‰¾å‡ºæœ€ä½³è°ƒåˆ¶è§£è°ƒå™¨
            best_modem = None
            if available_modems:
                # æŒ‰ä¿¡å·å¼ºåº¦å’ŒæˆåŠŸç‡æ’åº
                sorted_modems = sorted(
                    available_modems,
                    key=lambda m: (
                        int(m.get("signal_strength", 0)) if m.get("signal_strength", "0").isdigit() else 0,
                        m.get("success_rate", 0),
                        -m.get("last_used", 0)  # æœ€è¿‘ä½¿ç”¨æ—¶é—´è¶Šå°è¶Šå¥½
                    ),
                    reverse=True
                )
                best_modem = sorted_modems[0]

            # æ„å»ºè¯¦ç»†å¥åº·æ•°æ®
            health_data = {
                "timestamp": time.time(),
                "service_ready": connected and len(available_modems) > 0,
                "available_modems": len(available_modems),
                "busy_modems": len(busy_modems),
                "offline_modems": len(offline_modems),
                "total_modems": health_status["total_modems"],
                "best_modem": best_modem if best_modem else {},
                "available_modems_list": available_modems,
                "busy_modems_list": busy_modems,
                "offline_modems_list": offline_modems,
                "connection_test_passed": connected,
                "lock_timeout": self._lock_timeout,
                "send_timeout": self._send_timeout
            }

            # æ„å»ºçŠ¶æ€æ¶ˆæ¯
            if connected:
                if len(available_modems) > 0:
                    status_message = f"æœåŠ¡å¥åº· ({len(available_modems)}/{health_status['total_modems']} ä¸ªè°ƒåˆ¶è§£è°ƒå™¨å¯ç”¨"

                    if best_modem:
                        signal_str = best_modem.get("signal_strength", "N/A")
                        model = best_modem.get("model", "Unknown")
                        status_message += f"ï¼Œæœ€ä½³: {best_modem.get('port', 'N/A')} ({model}, ä¿¡å·: {signal_str})"

                    if len(busy_modems) > 0:
                        status_message += f"ï¼Œ{len(busy_modems)} ä¸ªå¿™ç¢Œä¸­"

                    return sms_pb2.HealthCheckResponse(
                        status=200,
                        message=status_message + ")",
                        data=json.dumps(health_data, ensure_ascii=False)
                    )
                else:
                    return sms_pb2.HealthCheckResponse(
                        status=503,
                        message=f"æœåŠ¡ä¸å¥åº·: 0/{health_status['total_modems']} ä¸ªè°ƒåˆ¶è§£è°ƒå™¨å¯ç”¨ï¼Œ{len(busy_modems)} ä¸ªå¿™ç¢Œä¸­",
                        data=json.dumps(health_data, ensure_ascii=False)
                    )
            else:
                return sms_pb2.HealthCheckResponse(
                    status=503,
                    message=f"æœåŠ¡ä¸å¥åº·: è¿æ¥æµ‹è¯•å¤±è´¥ï¼Œ{len(available_modems)}/{health_status['total_modems']} ä¸ªè°ƒåˆ¶è§£è°ƒå™¨å¯ç”¨",
                    data=json.dumps(health_data, ensure_ascii=False)
                )

        except Exception as e:
            logger.error(f"å¥åº·æ£€æŸ¥å¤±è´¥: {e}")
            error_data = {
                "timestamp": time.time(),
                "error": str(e),
                "service_ready": False,
                "details": "Health check failed with exception"
            }
            return sms_pb2.HealthCheckResponse(
                status=500,
                message=f"å¥åº·æ£€æŸ¥å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def GetSystemStatus(self, request, context) -> sms_pb2.HealthCheckResponse:
        """è·å–ç³»ç»ŸçŠ¶æ€ï¼ˆæ‰©å±•çš„å¥åº·æ£€æŸ¥ï¼‰"""
        try:
            # è°ƒç”¨æ ‡å‡†çš„å¥åº·æ£€æŸ¥
            health_response = await self.HealthCheck(request, context)

            # æ·»åŠ é¢å¤–çš„ç³»ç»Ÿä¿¡æ¯
            try:
                import asyncio
                import sys
                import platform
                import psutil

                health_data = json.loads(health_response.data)

                # æ·»åŠ ç³»ç»Ÿä¿¡æ¯
                health_data["system_info"] = {
                    "python_version": sys.version,
                    "platform": platform.platform(),
                    "asyncio_tasks": len(asyncio.all_tasks()),
                    "cpu_percent": psutil.cpu_percent(interval=0.1),
                    "memory_percent": psutil.virtual_memory().percent,
                    "disk_usage": psutil.disk_usage('/').percent
                }

                # æ›´æ–°å“åº”æ•°æ®
                health_response.data = json.dumps(health_data, ensure_ascii=False)

            except ImportError:
                # å¦‚æœpsutilä¸å¯ç”¨ï¼Œè·³è¿‡ç³»ç»Ÿä¿¡æ¯
                pass

            return health_response

        except Exception as e:
            logger.error(f"è·å–ç³»ç»ŸçŠ¶æ€å¤±è´¥: {e}")
            error_data = {
                "timestamp": time.time(),
                "error": str(e),
                "service_ready": False
            }
            return sms_pb2.HealthCheckResponse(
                status=500,
                message=f"è·å–ç³»ç»ŸçŠ¶æ€å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )
